core: add debug mode for mon and osd (#36)
rook/kubectl-rook-ceph: README.md

  - `status <CR>` : Print the phase and conditions of CRs of a specific type, such as `cephobjectstore`, `cephfilesystem`, etc
  - `purge-osd <osd-id> [--force]` : Permanently remove an OSD from the cluster. Multiple OSDs can be removed with a comma-separated list of IDs."

> - `debug <subcommand>...`           : Debug a deployment by scaling it down and creating a debug copy. This is supported for mons and OSDs only"

Reviewer: We really need more details documented about the behavior of this command.
- Describe the scenario when it could be used
- Explain that it will scale down the existing deployment, start a new debug deployment without probes and that sleeps forever so you can connect and run ceph commands

Perhaps we could put these details in a new Examples section below. We don't need to include the details here so we can keep this section simple.

Author: Done

  - `status <CR>` : Print the phase and conditions of CRs of a specific type, such as `cephobjectstore`, `cephfilesystem`, etc
  - `purge-osd <osd-id> [--force]` : Permanently remove an OSD from the cluster. Multiple OSDs can be removed with a comma-separated list of IDs.

> - `debug` : [Debug a deployment](#debug-mode)  by scaling it down and creating a debug copy. This is supported for mons and OSDs only"

Reviewer: ```suggestion
- `debug` : [Debug a deployment](#debug-mode)  by scaling it down and creating a debug copy. This is supported for mons and OSDs only
```

> - `debug <subcommand>...`           : Debug a deployment by scaling it down and creating a debug copy. This is supported for mons and OSDs only"

Reviewer: A link could help find the details
```suggestion
- `[debug](#debug-mode) <subcommand>...`           : Debug a deployment by scaling it down and creating a debug copy. This is supported for mons and OSDs only"
```

> - [debug](#debug-mode) `<subcommand>...`  : Debug a deployment by scaling it down and creating a debug copy. This is supported for mons and OSDs only"

Reviewer: The rendered link doesn't look quite right how I suggested it the first time. How about this?
```suggestion
- `debug <subcommand>...`  : [Debug a deployment](#debug-mode) by scaling it down and creating a debug copy. This is supported for mons and OSDs only"
```

> - `debug <subcommand>...`  : [Debug a deployment](#debug-mode)  by scaling it down and creating a debug copy. This is supported for mons and OSDs only"

Reviewer: I just realized the other commands don't mention subcommands in this doc. This seems consistent
```suggestion
- `debug`  : [Debug a deployment](#debug-mode)  by scaling it down and creating a debug copy. This is supported for mons and OSDs only"
```

Author: Yup

  - `start  <deployment-name> ` 

>     `[--alternate-image <alternate-image>]` : Start debugging a deployment with an optional container image"

Reviewer: ```suggestion
    `[--alternate-image <alternate-image>]` : Start debugging a deployment with an alternative ceph container image"
```

>     `[--alternate-image <alternate-image>]` : Start debugging a deployment with an optional alternative ceph container image"

Reviewer: ```suggestion
    `[--alternate-image <alternate-image>]` : Start debugging a deployment with an optional alternative ceph container image
```

Author: Thanks:)

}
```

### Debug Mode

> Debug Mode will be needed when cluster mon's or osd are been into error state or in the state when 

Reviewer: For lines 131-136:
```suggestion
Debug mode can be useful when a mon or OSD needs advanced maintenance operations that require the daemon to be stopped. Ceph tools such as `ceph-objectstore-tool`,`ceph-bluestore-tool`, or `ceph-monstore-tool` are commonly used in these scenarios. Debug mode will set up the mon or OSD so that these commands can be run.

Debug mode will automate the following:
1. Scale down the existing mon or OSD deployment
2. Start a new debug deployment where operations can be performed directly against the mon or OSD without that daemon running
   a. The main container sleeps so you can connect and run the ceph commands
   b. Liveness and startup probes are removed
```

the pods are not responding, So we use ceph tools like  ceph-objectstore-tool or ceph-bluestore-tool 
or ceph-monstore-tool, etc  for debugging.
Debug Mode will automate this process of using ceph tools for the opeartor. 

Debug Mode will work like it will scale down the existing deployment, start a new debug deployment without probes and that container pod will sleeps forever so you can connect and run ceph tools commands.

> Starting Debug Mode:

Reviewer: ```suggestion
For example, start the debug pod for mon `b`:
```

```console
kubectl rook-ceph debug start rook-ceph-mon-b
```
```text
setting debug mode for "rook-ceph-mon-b"
setting debug command to main container
deployment.apps/rook-ceph-mon-b scaled
deployment.apps/rook-ceph-mon-b-debug created
```

> 

Reviewer: Now connect to the daemon pod and perform operations:
```
kubectl exec <debug-pod> -- <ceph command>
```

> Stopping Debug Mode:

Reviewer: ```suggestion
When finished, stop debug mode and restore the original daemon:
```

   
   a. The main container sleeps so you can connect and run the ceph commands
   
   b. Liveness and startup probes are removed

   c. If alternate Image is passed by --alternate-image flag then the new debug deployment container will be using alternate Image.

For example, start the debug pod for mon `b`:
```console
kubectl rook-ceph debug start rook-ceph-mon-b
```
```text
setting debug mode for "rook-ceph-mon-b"
setting debug command to main container
deployment.apps/rook-ceph-mon-b scaled
deployment.apps/rook-ceph-mon-b-debug created
```

Now connect to the daemon pod and perform operations:
```console
kubectl exec <debug-pod> -- <ceph command>
```

When finished, stop debug mode and restore the original daemon:
```console
kubectl rook-ceph debug stop rook-ceph-mon-b
```
```text
setting debug mode for "rook-ceph-mon-b-debug"
removing debug mode from "rook-ceph-mon-b-debug"
deployment.apps "rook-ceph-mon-b-debug" deleted
deployment.apps/rook-ceph-mon-b scaled
```

>Note: If you need to update the limits and request of the debug deployment that is created using debug command you can run:

> >```console

Reviewer: The next section is rendered incorrectly, I think this character needs to be removed.
```suggestion
```console
```

Author: This is added intentionally,
And after this at cmd `> oc` also

So the console cmd can be the part of `Note` only

}
```

### Debug Mode

> Debug mode can be useful when a mon's or OSD needs advanced maintenance operations that require the daemon to be stopped. Ceph tools such as `ceph-objectstore-tool`,`ceph-bluestore-tool`, or `ceph-monstore-tool` are commonly used in these scenarios. Debug mode will set up the mon or OSD so that these commands can be run.

Reviewer: ```suggestion
Debug mode can be useful when a mon or OSD needs advanced maintenance operations that require the daemon to be stopped. Ceph tools such as `ceph-objectstore-tool`,`ceph-bluestore-tool`, or `ceph-monstore-tool` are commonly used in these scenarios. Debug mode will set up the mon or OSD so that these commands can be run.
```

Debug mode will automate the following:
1. Scale down the existing mon or OSD deployment
2. Start a new debug deployment where operations can be performed directly against the mon or OSD without that daemon running
   
   a. The main container sleeps so you can connect and run the ceph commands
   
   b. Liveness and startup probes are removed

   c. If alternate Image is passed by --alternate-image flag then the new debug deployment container will be using alternate Image.

> Debug Mode will work like it will scale down the existing deployment, start a new debug deployment without probes and that container pod will sleeps forever so you can connect and run ceph tools commands.

Reviewer: We don't need this line anymore, right? It's covered by the bullets above

For example, start the debug pod for mon `b`:
```console
kubectl rook-ceph debug start rook-ceph-mon-b
```
```text
setting debug mode for "rook-ceph-mon-b"
setting debug command to main container
deployment.apps/rook-ceph-mon-b scaled
deployment.apps/rook-ceph-mon-b-debug created
```

Now connect to the daemon pod and perform operations:
```console
kubectl exec <debug-pod> -- <ceph command>
```

> You can also use alternate-image while starting the debug mode:

Reviewer: I'd suggest removing lines 161-164, the alternate-image flag seems self explanatory. Let's keep this example focused.
